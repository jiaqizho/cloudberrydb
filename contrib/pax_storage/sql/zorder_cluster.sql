-- test cluster_columns
set pax_max_tuples_per_file to 131072;
drop table if EXISTS t_zorder_cluster;
create table t_zorder_cluster(c1 int, c2 int) using pax with(minmax_columns='c1,c2');
\d+ t_zorder_cluster;

insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;

select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_zorder_cluster');

-- error, becuase no cluster index or cluster_columns is defined
cluster t_zorder_cluster;

alter table t_zorder_cluster set(cluster_columns='c1,c2');
\d+ t_zorder_cluster;
-- success
cluster t_zorder_cluster;
select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_zorder_cluster');


drop table t_zorder_cluster;

-- test cluster index
set pax_max_tuples_per_file to 131072;
drop table if EXISTS t_index_cluster;
create table t_index_cluster(c1 int, c2 int) using pax with(minmax_columns='c1,c2');
\d+ t_index_cluster;

insert into t_index_cluster select i,i from generate_series(1,100000) i;
insert into t_index_cluster select i,i from generate_series(1,100000) i;
insert into t_index_cluster select i,i from generate_series(1,100000) i;
insert into t_index_cluster select i,i from generate_series(1,100000) i;
insert into t_index_cluster select i,i from generate_series(1,100000) i;

select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_index_cluster');

-- error, becuase no cluster index or cluster_columns is defined
cluster t_index_cluster;

create index idx_t_index_cluster on t_index_cluster(c1);
\d+ t_index_cluster;

-- success
cluster t_index_cluster using idx_t_index_cluster;

select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_index_cluster');
drop table t_index_cluster;

-- test both cluster index and cluster columns
create table t_both_zorder_and_index_cluster(c1 int, c2 int) using pax with(minmax_columns='c1,c2');
alter table t_both_zorder_and_index_cluster set(cluster_columns='c1,c2');
create index idx_t_both_zorder_and_index_cluster on t_both_zorder_and_index_cluster(c1);
-- error, because zorder cluster and index cluster can not be used together
cluster t_both_zorder_and_index_cluster using idx_t_both_zorder_and_index_cluster;
-- success
cluster t_both_zorder_and_index_cluster;
alter table t_both_zorder_and_index_cluster set(cluster_columns='');
-- error, because zorder cluster reloptions has been deleted
cluster t_both_zorder_and_index_cluster;
-- success
cluster t_both_zorder_and_index_cluster using idx_t_both_zorder_and_index_cluster;
-- error
alter table t_both_zorder_and_index_cluster set(cluster_columns='c1,c2');
drop index idx_t_both_zorder_and_index_cluster;
-- error
cluster t_both_zorder_and_index_cluster;
-- success
alter table t_both_zorder_and_index_cluster set(cluster_columns='c1,c2');
-- success
cluster t_both_zorder_and_index_cluster;

-- test unsupport type

create table t_zorder_unsupport_type(c1 int, c2 numeric(10,2),c3 varchar(128), c4 timestamp, c5 bpchar(64) ) using pax with(minmax_columns='c1,c2');

-- error, because numeric is unsupport type
alter table t_zorder_unsupport_type set(cluster_columns='c1,c2');
alter table t_zorder_unsupport_type set(cluster_columns='c1,c3');
-- error, because timestamp is unsupport type
alter table t_zorder_unsupport_type set(cluster_columns='c1,c4');
alter table t_zorder_unsupport_type set(cluster_columns='c1,c5');
