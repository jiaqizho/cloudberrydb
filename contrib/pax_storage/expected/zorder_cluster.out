-- test cluster_columns
set pax_max_tuples_per_file to 131072;
drop table if EXISTS t_zorder_cluster;
create table t_zorder_cluster(c1 int, c2 int) using pax with(minmax_columns='c1,c2');
\d+ t_zorder_cluster;
                             Table "public.t_zorder_cluster"
 Column |  Type   | Collation | Nullable | Default | Storage | Stats target | Description 
--------+---------+-----------+----------+---------+---------+--------------+-------------
 c1     | integer |           |          |         | plain   |              | 
 c2     | integer |           |          |         | plain   |              | 
Distributed by: (c1)
Options: minmax_columns=c1,c2

insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_zorder_cluster');
 ptblockname |                                          ptstatistics                                           | ptisclustered 
-------------+-------------------------------------------------------------------------------------------------+---------------
           0 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           1 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           2 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           3 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           4 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           0 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
           1 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
           2 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
           3 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
           4 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
           0 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
           1 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
           2 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
           3 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
           4 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
(15 rows)

-- error, becuase no cluster index or cluster_columns is defined
cluster t_zorder_cluster;
ERROR:  there is no previously clustered index or cluster_columns reloptions for table "t_zorder_cluster"
alter table t_zorder_cluster set(cluster_columns='c1,c2');
\d+ t_zorder_cluster;
                             Table "public.t_zorder_cluster"
 Column |  Type   | Collation | Nullable | Default | Storage | Stats target | Description 
--------+---------+-----------+----------+---------+---------+--------------+-------------
 c1     | integer |           |          |         | plain   |              | 
 c2     | integer |           |          |         | plain   |              | 
Distributed by: (c1)
Options: minmax_columns=c1,c2, cluster_columns=c1,c2

-- success
cluster t_zorder_cluster;
select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_zorder_cluster');
 ptblockname |                                              ptstatistics                                               | ptisclustered 
-------------+---------------------------------------------------------------------------------------------------------+---------------
           5 | [(false,false),(131072),(1,78633),(5126644321)],[(false,false),(131072),(1,78633),(5126644321)]         | t
           6 | [(false,false),(35563),(78633,99996),(3178950384)],[(false,false),(35563),(78633,99996),(3178950384)]   | t
           5 | [(false,false),(131072),(2,78230),(5147801190)],[(false,false),(131072),(2,78230),(5147801190)]         | t
           6 | [(false,false),(36238),(78230,99999),(3226202455)],[(false,false),(36238),(78230,99999),(3226202455)]   | t
           5 | [(false,false),(131072),(5,79072),(5187913809)],[(false,false),(131072),(5,79072),(5187913809)]         | t
           6 | [(false,false),(34983),(79072,100000),(3132737841)],[(false,false),(34983),(79072,100000),(3132737841)] | t
(6 rows)

drop table t_zorder_cluster;
-- test cluster index
set pax_max_tuples_per_file to 131072;
drop table if EXISTS t_index_cluster;
NOTICE:  table "t_index_cluster" does not exist, skipping
create table t_index_cluster(c1 int, c2 int) using pax with(minmax_columns='c1,c2');
\d+ t_index_cluster;
                              Table "public.t_index_cluster"
 Column |  Type   | Collation | Nullable | Default | Storage | Stats target | Description 
--------+---------+-----------+----------+---------+---------+--------------+-------------
 c1     | integer |           |          |         | plain   |              | 
 c2     | integer |           |          |         | plain   |              | 
Distributed by: (c1)
Options: minmax_columns=c1,c2

insert into t_index_cluster select i,i from generate_series(1,100000) i;
insert into t_index_cluster select i,i from generate_series(1,100000) i;
insert into t_index_cluster select i,i from generate_series(1,100000) i;
insert into t_index_cluster select i,i from generate_series(1,100000) i;
insert into t_index_cluster select i,i from generate_series(1,100000) i;
select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_index_cluster');
 ptblockname |                                          ptstatistics                                           | ptisclustered 
-------------+-------------------------------------------------------------------------------------------------+---------------
           0 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
           1 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
           2 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
           3 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
           4 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
           0 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           1 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           2 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           3 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           4 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           0 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
           1 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
           2 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
           3 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
           4 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
(15 rows)

-- error, becuase no cluster index or cluster_columns is defined
cluster t_index_cluster;
ERROR:  there is no previously clustered index or cluster_columns reloptions for table "t_index_cluster"
create index idx_t_index_cluster on t_index_cluster(c1);
\d+ t_index_cluster;
                              Table "public.t_index_cluster"
 Column |  Type   | Collation | Nullable | Default | Storage | Stats target | Description 
--------+---------+-----------+----------+---------+---------+--------------+-------------
 c1     | integer |           |          |         | plain   |              | 
 c2     | integer |           |          |         | plain   |              | 
Indexes:
    "idx_t_index_cluster" btree (c1)
Distributed by: (c1)
Options: minmax_columns=c1,c2

-- success
cluster t_index_cluster using idx_t_index_cluster;
select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_index_cluster');
 ptblockname |                                              ptstatistics                                               | ptisclustered 
-------------+---------------------------------------------------------------------------------------------------------+---------------
           0 | [(false,false),(131072),(2,78230),(5147801190)],[(false,false),(131072),(2,78230),(5147801190)]         | t
           1 | [(false,false),(36238),(78230,99999),(3226202455)],[(false,false),(36238),(78230,99999),(3226202455)]   | t
           0 | [(false,false),(131072),(1,78633),(5126644321)],[(false,false),(131072),(1,78633),(5126644321)]         | t
           1 | [(false,false),(35563),(78633,99996),(3178950384)],[(false,false),(35563),(78633,99996),(3178950384)]   | t
           0 | [(false,false),(131072),(5,79072),(5187913809)],[(false,false),(131072),(5,79072),(5187913809)]         | t
           1 | [(false,false),(34983),(79072,100000),(3132737841)],[(false,false),(34983),(79072,100000),(3132737841)] | t
(6 rows)

drop table t_index_cluster;
-- test both cluster index and cluster columns
create table t_both_zorder_and_index_cluster(c1 int, c2 int) using pax with(minmax_columns='c1,c2');
alter table t_both_zorder_and_index_cluster set(cluster_columns='c1,c2');
create index idx_t_both_zorder_and_index_cluster on t_both_zorder_and_index_cluster(c1);
-- error, because zorder cluster and index cluster can not be used together
cluster t_both_zorder_and_index_cluster using idx_t_both_zorder_and_index_cluster;
ERROR:  cannot using index to cluster table which has cluster-columns (pax_access_handle.cc:1186)
-- success
cluster t_both_zorder_and_index_cluster;
alter table t_both_zorder_and_index_cluster set(cluster_columns='');
-- error, because zorder cluster reloptions has been deleted
cluster t_both_zorder_and_index_cluster;
ERROR:  there is no previously clustered index or cluster_columns reloptions for table "t_both_zorder_and_index_cluster"
-- success
cluster t_both_zorder_and_index_cluster using idx_t_both_zorder_and_index_cluster;
-- error
alter table t_both_zorder_and_index_cluster set(cluster_columns='c1,c2');
ERROR:  pax table has previously clustered index, can't set zorder cluster reloptions (pax_access_handle.cc:1308)
drop index idx_t_both_zorder_and_index_cluster;
-- error
cluster t_both_zorder_and_index_cluster;
ERROR:  there is no previously clustered index or cluster_columns reloptions for table "t_both_zorder_and_index_cluster"
-- success
alter table t_both_zorder_and_index_cluster set(cluster_columns='c1,c2');
-- success
cluster t_both_zorder_and_index_cluster;
-- test unsupport type
create table t_zorder_unsupport_type(c1 int, c2 numeric(10,2),c3 varchar(128), c4 timestamp, c5 bpchar(64) ) using pax with(minmax_columns='c1,c2');
-- error, because numeric is unsupport type
alter table t_zorder_unsupport_type set(cluster_columns='c1,c2');
ERROR:  the type of column c2 does not support zorder cluster (paxc_rel_options.cc:312)
alter table t_zorder_unsupport_type set(cluster_columns='c1,c3');
-- error, because timestamp is unsupport type
alter table t_zorder_unsupport_type set(cluster_columns='c1,c4');
ERROR:  the type of column c4 does not support zorder cluster (paxc_rel_options.cc:312)
alter table t_zorder_unsupport_type set(cluster_columns='c1,c5');
